### PR TITLE
Change requires_grad to Optional[bool] in torch.asarray

### LINKED ISSUE #170819: `asarray` removes `requires_grad`
### 🐛 Describe the bug

Torch strips tensors of the `requires_grad` attribute when passing them through `asarray`. This is problematic for downsteam libraries like scipy, which use it to canonize inputs. As a consequence, users cannot differentiate through these functions.

Now _technically_ the standard does not specify that the resulting tensor should keep its attributes, and `requires_grad` is thus not covered by it. However, it does specify that the `requires_grad` keyword argument should not exist in the first place: https://data-apis.org/array-api/latest/API_specification/generated/array_api.asarray.html

Irrespective of whether strict adherence to the standard is desirable, the sequence below is highly unintuitive for users / borderline buggy.

```python
import torch
q = torch.tensor([0.0, 0.0, 0.0, 1.0], requires_grad=True)
print(q.requires_grad)  # True

q2 = torch.asarray(q)
print(q2.requires_grad)  # False
print(q.requires_grad)  # False
print(q is q2)  # True
```
The best way of working around this issue while keeping the `requires_grad` attribute would be to make it `bool | None`. If `None`, the flag is copied from the tensor. Arrays from external frameworks without the `requires_grad` flag default to `False`. The default value for the `requires_grad` argument of `asarray` could then be changed to `None`, producing the expected/desired behaviour.

See https://github.com/scipy/scipy/issues/24186 for an example where this becomes a real issue.

### Versions

PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (conda-forge gcc 13.3.0-2) 13.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.13.2 | packaged by conda-forge | (main, Feb 17 2025, 14:10:22) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.6.87.2-microsoft-standard-WSL2-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 581.42
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        48 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               32
On-line CPU(s) list:                  0-31
Vendor ID:                            AuthenticAMD
Model name:                           AMD Ryzen 9 7950X 16-Core Processor
CPU family:                           25
Model:                                97
Thread(s) per core:                   2
Core(s) per socket:                   16
Socket(s):                            1
Stepping:                             2
BogoMIPS:                             8982.99
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl tsc_reliable nonstop_tsc cpuid extd_apicid tsc_known_freq pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw topoext perfctr_core ssbd ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512_bf16 clzero xsaveerptr arat npt nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload avx512vbmi umip avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid fsrm
Virtualization:                       AMD-V
Hypervisor vendor:                    Microsoft
Virtualization type:                  full
L1d cache:                            512 KiB (16 instances)
L1i cache:                            512 KiB (16 instances)
L2 cache:                             16 MiB (16 instances)
L3 cache:                             32 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-31
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Vulnerable: Safe RET, no microcode
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; IBPB conditional; IBRS_FW; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] mypy==1.15.0
[pip3] mypy_extensions==1.0.0
[pip3] numpy==2.2.4
[pip3] numpydoc==1.8.0
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] torch==2.6.0
[pip3] triton==3.2.0
[conda] numpy                     2.2.4           py313h17eae1a_0    conda-forge
[conda] numpydoc                  1.8.0              pyhd8ed1ab_1    conda-forge
[conda] nvidia-cublas-cu12        12.4.5.8                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.4.127                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.2.1.3                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.5.147               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.6.1.9                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.3.1.170               pypi_0    pypi
[conda] nvidia-cusparselt-cu12    0.6.2                    pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.4.127                 pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.4.127                 pypi_0    pypi
[conda] torch                     2.6.0                    pypi_0    pypi
[conda] triton                    3.2.0                    pypi_0    pypi

cc @mruberry @rgommers @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi

### PR DESCRIPTION
**Pull request by benjaminglass1**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #170897

The current default behavior when invoking `torch.asarray` is to drop any existing gradients on the input. This is confusing when inputting a tensor object, so this commit changes the default to follow the input iff the input is a tensor (while preserving explicit override behavior). This is technically an API breaking change, but a minor one that I suspect will not affect downstream users.

Fixes #170819.

cc @mruberry @rgommers @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi

### CODE REVIEW

File: torch/csrc/utils/tensor_new.cpp [hunk 1]
```diff
@@ -1756,9 +1756,11 @@ Tensor asarray(
   if (THPVariable_Check(obj)) {
     tensor = THPVariable_Unpack(obj);
   }
+  bool return_requires_grad =
```
Comment (amjames) [inline torch/csrc/utils/tensor_new.cpp:1764]: If this is going to be considered an acceptable BC break this would be an appropriate place for the warning to fire according to [policy](https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy#minimizing-the-disruption-of-bc-breaking-changes)

**Review by amjames — COMMENTED** [review]
I do agree that the new behavior would be less surprising, but it is backward incompatible. 

It is common to branch on the value of `requires_grad` for a variety of reasons, and this change will cause those points to change from one path to another. How common is it to use this API with input tensors that require_grad and then later pass through a check on the `requires_grad` value of the output, or one of the downstream outputs I don't know. 

Others will have to weigh in on how potentially disruptive this might be. I am in favor of the change but we will need to warn_once (see in-line comment) at the very least.

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (benjaminglass1) [reply torch/csrc/utils/tensor_new.cpp:1764]: Ahh good to know, I'll add the warning.

Comment (benjaminglass1) [reply torch/csrc/utils/tensor_new.cpp:1764]: Warning added, opening up this PR for review.

**Review by soulitzer — APPROVED** [review]
Thanks!

File: torch/csrc/utils/tensor_new.cpp [hunk 2]
```diff
@@ -1761,9 +1761,15 @@ Tensor asarray(
...
+      requires_grad.value_or(tensor.defined() ? tensor.requires_grad() : false);
+  if (return_requires_grad && !requires_grad) {
+    TORCH_WARN_ONCE(
+        "Unspecified requires_grad now defaults to obj.requires_grad, not False!")
```
Comment (soulitzer) [inline torch/csrc/utils/tensor_new.cpp]: Let's make the warning something like:
```
  TORCH_WARN_ONCE(
      "torch.asarray: unspecified requires_grad now defaults to "
      "obj.requires_grad instead of False. Pass requires_grad=False "
      "explicitly to get the old behavior and silence this warning.");
```

Comment (soulitzer) [reply torch/csrc/utils/tensor_new.cpp]: Is the plan to remove this warning in a future release?

Comment (benjaminglass1) [reply torch/csrc/utils/tensor_new.cpp]: @soulitzer Yes, this should be removed in a future release. I suspect that very few users will encounter this warning and prefer the old behavior (it's very counter-intuitive imo), but I will update as suggested then merge!
